File: findClosestElements.py
import collections
import bisect
#Approach 1- Custom sorting (brute force)
def findClosestElements(arr,k,x): # TC O(n log n + k log k) // SC O(n)
	sorted_arr=sorted(arr, key=lambda nums:abs(x-nums))

	result=sorted_arr[:k]

	return sorted(result)

# Approach 2- Binary Search + Sliding window 
def findClosestElements1(arr,k,x): # TC O(log n + k) // SC O(1)
	if len(arr)==k: return arr 

	left=bisect.bisect_left(arr,x)-1
	right=left+1 

	while right-left-1<k:
		if left==-1:
			right+=1 

		if right==len(arr) or abs(arr[left]-x)<=abs(arr[right]-x):
			left-=1
		else:
			right+=1 


	return arr[left+1:right]

#Approach 3- Binary search to find the left bound
def findClosestElements2(arr,k,x): # TC O(log(n-k)+k) // SC O(1) 
    # Initialize binary search bounds
    left = 0
    right = len(arr) - k
    
    # Binary search against the criteria described
    while left < right:
        mid = (left + right) // 2
        if x - arr[mid] > arr[mid + k] - x:
            left = mid + 1
        else:
            right = mid

    return arr[left:left + k]

print(findClosestElements([1,2,3,4,5],4,3))
print(findClosestElements1([1,2,3,4,5],4,3))
print(findClosestElements2([1,2,3,4,5],4,3))

File: findClosestElements.java
import java.util.*;
// Approach 1- Custom Sorting (Brute force)
class Solution { // TC O(n log n + k log k) SC O(n)
    public List<Integer> findClosestElements(int[] arr, int k, int x) {
        // Convert from array to list first to make use of Collections.sort()
        List<Integer> sortedArr = new ArrayList<Integer>();
        for (int num: arr) {
            sortedArr.add(num);
        }
        
        // Sort using custom comparator
        Collections.sort(sortedArr, (num1, num2) -> Math.abs(num1 - x) - Math.abs(num2 - x));
        
        // Only take k elements
        sortedArr = sortedArr.subList(0, k);
        
        // Sort again to have output in ascending order
        Collections.sort(sortedArr);
        return sortedArr;
    }
}

// Approach 2- Binary Search + sliding window 
class Solution1{ // TC O(log n + k) // SC O(1)
    public List<Integer> findClosestElements(int[] arr, int k, int x) {
        List<Integer> result = new ArrayList<Integer>();
        
        // Base case
        if (arr.length == k) {
            for (int i = 0; i < k; i++) {
                result.add(arr[i]);
            }
            
            return result;
        }
        
        // Binary search to find the closest element
        int left = 0;
        int right = arr.length;
        int mid = 0;
        while (left < right) {
            mid = (left + right) / 2;
            if (arr[mid] >= x) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        
        // Initialize our sliding window's bounds
        left -= 1;
        right = left + 1;
        
        // While the window size is less than k
        while (right - left - 1 < k) {
            // Be careful to not go out of bounds
            if (left == -1) {
                right += 1;
                continue;
            }
            
            // Expand the window towards the side with the closer number
            // Be careful to not go out of bounds with the pointers
            if (right == arr.length || Math.abs(arr[left] - x) <= Math.abs(arr[right] - x)) {
                left -= 1;
            } else {
                right += 1;
            }
        } 

        // Build and return the window
        for (int i = left + 1; i < right; i++) {
            result.add(arr[i]);
        }
        
        return result;
    }
}

// Approach 3- Binary Search to find the left bound 
class Solution2{ // TC O(log (n-k)+ k) SC O(1)
    public List<Integer> findClosestElements(int[] arr, int k, int x) {
        // Initialize binary search bounds
        int left = 0;
        int right = arr.length - k;
        
        // Binary search against the criteria described
        while (left < right) {
            int mid = (left + right) / 2;
            if (x - arr[mid] > arr[mid + k] - x) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        
        // Create output in correct format
        List<Integer> result = new ArrayList<Integer>();
        for (int i = left; i < left + k; i++) {
            result.add(arr[i]);
        }
        
        return result;
    }
}


class findClosestElements{
    public static void main(String[] args) {
        int[] arr={1,2,3,4,5};
        int k=4;
        int x=3;
        Solution s= new Solution();
        System.out.println(s.findClosestElements(arr,k,x));
        Solution1 s1= new Solution1();
        System.out.println(s1.findClosestElements(arr,k,x));
        Solution2 s2= new Solution2();
        System.out.println(s2.findClosestElements(arr,k,x));
    }
}